### ISSUE TITLE
Implement AOPP Support

### ISSUE THREAD

**Issue by lclc** [sparrowwallet/sparrow#113]
AOPP (Address Ownership Proof Protocol) is a new, simple protocol that specifies the transmission of signed messages via a URI scheme: https://gitlab.com/aopp/address-ownership-proof-protocol

Providing proof of ownership over an address is required to withdrawal Bitcoin in some countries (Switzerland, the Netherlands, Singapore, and soon more). It's a hassle to do that manually for the user and the exchange. AOPP heavily improves that. See [aopp.group](aopp.group) for more information about it.

Besides improving the UX for the user, there are two more benefits:

1. There is no need to copy & paste a Bitcoin address for the wallet to the exchange to withdraw. This prevents malware from replacing it in the clipboard (e.g. https://www.fortinet.com/blog/threat-research/vipersoftx-new-javascript-threat)
2. The wallet is aware that the user is withdrawing from an exchange (even which exchange) and with that can suggest other options, which might be better for his privacy, provide better fees or better technology (e.g. segwit support, etc.).

AOPP does not affect and is not even visible to users of exchanges who don't have AOPP (don't require ownership proof).

As Sparrow already supports URI handling for _bitcoin:_ and the signing of messages, this shouldn't be too hard to implement. BlueWallet (which also has both) implement AOPP support in less than a day (https://github.com/BlueWallet/BlueWallet/pull/2915).

**Comment by craigraw** [comment]
Implemented in 425e476. The OS registration for the `aopp:` URI handling still needs to be tested on Linux and Windows installs.

**Comment by craigraw** [comment]
A note on the server implementation: Sparrow follows the 'Electrum signing' approach as directed when performing the signing with private keys it has access to. However it should be noted that when signing with a hardware wallet, this parameter cannot be specified. There is no reason however why the AOPP server cannot verify both the Electrum and the Trezor (BIP137) type signatures, and produce a successful response on either. Sparrow follows this approach when verifying signatures itself.

**Comment by dspicher** [comment]
> Implemented in [425e476](https://github.com/sparrowwallet/sparrow/commit/425e476f202273dcc4e721a3aa92defca31196f4). The OS registration for the `aopp:` URI handling still needs to be tested on Linux and Windows installs.

Amazing news, that was very quick!

I quickly tested 425e476 on Ubuntu 20.04 LTS and I'm indeed running into the following when trying to start the app:
```
java.lang.reflect.InvocationTargetException
...
Caused by: java.lang.UnsupportedOperationException: The APP_OPEN_URI action is not supported on the current platform!
        at java.desktop/java.awt.Desktop.checkActionSupport(Desktop.java:380)
        at java.desktop/java.awt.Desktop.setOpenURIHandler(Desktop.java:814)
        at com.sparrowwallet.sparrow@1.4.0/com.sparrowwallet.sparrow.AppServices.addURIHandlers(AppServices.java:604)
        at com.sparrowwallet.sparrow@1.4.0/com.sparrowwallet.sparrow.AppServices.start(AppServices.java:146)
        at com.sparrowwallet.sparrow@1.4.0/com.sparrowwallet.sparrow.MainApp.start(MainApp.java:111)
...
```

**Comment by toxeus** [comment]
> when signing with a hardware wallet, this parameter cannot be specified

AFAIK all hardware wallets support 'Electrum signing' (should be called 'Bitcoin Core signing' IMO) and some support both schemes. 

> There is no reason however why the AOPP server cannot verify both 

It's a brand new protocol so there's no good reason to inherit the inconsistencies from the traditional message singing where indeed both signature schemes need to be supported unfortunately.

**Comment by craigraw** [comment]
> AFAIK all hardware wallets support 'Electrum signing' (should be called 'Bitcoin Core signing' IMO) and some support both schemes.

Have Trezor switched away from BIP137? Even if they have, there will be plenty of old firmware devices out there. Seems unlikely they have though - can you link to this change of theirs in Github? Would be good to understand their rationale.

AFAIK Bitcoin Core refuses to sign messages with non-P2PKH addresses, so the 'Electrum' approach is due to Electrum's interpretation of how to take things forward with Segwit addresses. It's a mess, but we can verify both easily enough.

**Comment by craigraw** [comment]
> I've added the Sparrow logo to aopp.group. I hope that's OK, otherwise let me know.

All good, thank you.

**Comment by toxeus** [comment]
> Have Trezor switched away from BIP137?

No, they support both to be compatible with Bitcoin Core. I'm currently on mobile but will try to provide relevant links when once I have a real device at my disposal.

**Comment by craigraw** [comment]
Now working on OSX, Linux and Windows. The OS behaviour on Linux and Windows is to open a new instance of Sparrow to perform the URI action, which is not always ideal but works well enough in practice.

On OSX, the passing the URI through for handling only occurs when Sparrow is already running and has registered the URI handler with the operating system. However, clicking an `aopp:` link will at least open Sparrow if it is not running, so a second click on the link will perform the proof even in this case.

> No, they support both to be compatible with Bitcoin Core.

I'm unsure how this is done, since there is no option on the Trezor to select which format of signature to create. However I have tested with a Trezor One on a wrapped segwit address and it works. I haven't investigated further yet.

Closing this off for now as implemented.

**Comment by RubenWaterman** [comment]
I've tried this on my website (getbittr.com) but the Sparrow implementation doesn't seem to work. In the first version (I think) with AOPP enabled (1.4.1), Sparrow was sending an address and version number to our server but not the signature so the validation didn't work. In the latest version (1.4.3), Sparrow doesn't send anything at all after clicking "Sign and send".

Also testing the AOPP demo doesn't work with Sparrow ([https://testing.aopp.group/?&msg=vasp-chosen-msg&asset=btc&format=any](https://testing.aopp.group/?&msg=vasp-chosen-msg&asset=btc&format=any)) so I don't think the issue is on getbittr's end.

**Comment by craigraw** [comment]
This was indeed a regression. Fixed in 34d1571. 

I have just successfully registered an address on testing.aopp.group. Thanks for the report.
